sw/device/lib/crypto/impl/aes.c

```diff
@@ -0,0 +1,325 @@
+  if (partial_data_len >= kAesBlockNumBytes) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Get a byte-sized pointer to the padding start point within the block's
+  // data buffer.
+  char *padding = ((char *)block->data) + partial_data_len;
+
+  // Pad a full block if the last block is full, otherwise just fill the last
+  // block.
+  size_t padding_len = kAesBlockNumBytes - partial_data_len;
+  hardened_bool_t padding_written = kHardenedBoolFalse;
+  switch (launder32(padding_mode)) {
+    case kAesPaddingPkcs7:
+      // Pads with value same as the number of padding bytes.
+      memset(padding, padding_len, padding_len);
+      padding_written = kHardenedBoolTrue;
+      break;
+    case kAesPaddingIso9797M2:
+      // Pads with 0x80 (0b10000000), followed by zero bytes.
+      memset(padding, 0x0, padding_len);
+      padding[0] = 0x80;
+      padding_written = kHardenedBoolTrue;
+      break;
+    case kAesPaddingNull:
+      // This routine should not be called if padding is not needed.
+      return OTCRYPTO_BAD_ARGS;
+    default:
+      // TODO(#15591): Add any other padding modes that will be included in the
+      // final API.
+      return OTCRYPTO_BAD_ARGS;
+  }
+  HARDENED_CHECK_EQ(padding_written, kHardenedBoolTrue);
+
+  return OTCRYPTO_OK;
+}
+
+/**
+ * Return the number of blocks for the input and padding mode.
+ *
+ * @param plaintext_len Length of plaintext in bytes.
+ * @param padding Padding mode.
+ * @returns Number of AES blocks required.
+ */
+static status_t num_padded_blocks_get(size_t plaintext_len,
+                                      aes_padding_t padding,
+                                      size_t *num_blocks) {
+  size_t num_full_blocks = plaintext_len / kAesBlockNumBytes;
+
+  if (padding == kAesPaddingNull) {
+    // If no padding mode is given, the last block must be full.
+    if (num_full_blocks * kAesBlockNumBytes != plaintext_len) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+
+    *num_blocks = num_full_blocks;
+    return OTCRYPTO_OK;
+  }
+  HARDENED_CHECK_NE(padding, kAesPaddingNull);
+
+  // For non-null padding modes, we append a full block of padding if the last
+  // block is full, so the value is always <full blocks> + 1.
+  *num_blocks = num_full_blocks + 1;
+  return OTCRYPTO_OK;
+}
+
+/**
+ * Return the block at index i for the given input and padding mode.
+ *
+ * @param input Input data buffer.
+ * @param padding Padding mode.
+ * @param index Block index.
+ * @param[out] AES block at `index`.
+ * @param input_len Length of cipher input in bytes.
+ * @param padding Padding mode.
+ * @returns Number of AES blocks required.
+ */
+static status_t get_block(crypto_const_uint8_buf_t input, aes_padding_t padding,
+                          size_t index, aes_block_t *block) {
+  size_t num_full_blocks = input.len / kAesBlockNumBytes;
+
+  // The index should never be more than `num_full_blocks` + 1, even including
+  // padding.
+  if (index > num_full_blocks + 1) {
```

Comment:
I can't remember if even in this case we want to launder the value to make sure the comparison doesn't get optimized out when we turn on LTO. This applies to other places as well. 

`HARDENED_CHECK` also launders, but I am not sure if that is sufficient. 

```c
if (launder32(index) > num_full_blocks + 1)
```

```

Reply:
I've changed this to just a `HARDENED_CHECK_LE`, since it's an internal function and should never get called with an out-of-range index in normal operation. I checked the assembly and it's not optimized away.